### PR TITLE
feat(sc): add support to map parameter

### PR DESCRIPTION
**Pull request by raulduartep**

### CODE REVIEW

**Comment by lllwvlvwlll** [comment]
Task linked: [CU-864ej5xh6 On ContractParam of neon-js, implement Map type](https://app.clickup.com/t/864ej5xh6)

File: packages/neon-core/src/sc/ContractParam.ts [hunk 1]
```diff
@@ -435,18 +501,31 @@ export class ContractParam implements NeonObject<ContractParamLike> {
...
           }
           return false;
 
+        case ContractParamType.Map:
+          if (
+            Array.isArray(this.value) &&
+            Array.isArray(other.value) &&
+            this.value.length === other.value.length
+          ) {
+            return (this.value as ContractParamMap).every(
```
Comment (snowypowers) [inline packages/neon-core/src/sc/ContractParam.ts:521]: this assumes that the map is ordered. Is this the case? (u have no tests for this at the moment)

File: packages/neon-core/src/sc/ContractParam.ts [hunk 2]
```diff
@@ -16,16 +16,31 @@ export enum ContractParamType {
...
 
 export interface ContractParamJson {
   type: string;
-  value?: string | boolean | number | ContractParamJson[] | null;
+  value?:
+    | string
+    | boolean
+    | number
+    | ContractParamJson[]
+    | ContractParamMapLike
```
Comment (snowypowers) [inline packages/neon-core/src/sc/ContractParam.ts:42]: The json representation should only include json stuff. So you will need a `ContractParamMapJson` type here. (The usage of the *Json types is at the serialization boundary so we should be assured that the input is always plain json)

File: packages/neon-core/src/sc/ContractParam.ts [hunk 3]
```diff
@@ -16,16 +16,31 @@ export enum ContractParamType {
...
 
   Void = 0xff,
 }
+export type ContractParamMapLike = {
+  key: ContractParamLike;
+  value: ContractParamLike;
+}[];
+
+export type ContractParamMap = {
+  key: ContractParam;
```
Comment (snowypowers) [inline packages/neon-core/src/sc/ContractParam.ts:31]: on a side note, does neo actually allow anything to be the key? Can I put a map as the key?

Comment (ixje) [reply packages/neon-core/src/sc/ContractParam.ts:31]: No you can't. Only `string`, `bool` and `int` can be used as key type (what they call `PrimitiveType` in the VM).

File: packages/neon-core/src/sc/ContractParam.ts (hunk 1 above)
Comment (ixje) [reply packages/neon-core/src/sc/ContractParam.ts:521]: a `map` on the VM side is ordered [ref](https://github.com/neo-project/neo-vm/blob/5b0a39811b34abacab1273f3ee5a9a9f7e52ac7b/src/Neo.VM/Types/Map.cs#L21)

File: packages/neon-core/src/sc/ContractParam.ts (hunk 3 above)
Comment (raulduartep) [reply packages/neon-core/src/sc/ContractParam.ts:31]: @snowypowers Should I create a validation for it ?

Comment (snowypowers) [reply packages/neon-core/src/sc/ContractParam.ts:31]: If we can constraint the type, then yes please. Else, at least some comments would help out.

File: packages/neon-core/src/sc/ContractParam.ts [hunk 4]
```diff
@@ -16,22 +16,44 @@ export enum ContractParamType {
...
   Array = 0x20,
-  Map = 0x22, // TODO: Implement support
+  Map = 0x22,
 
   InteropInterface = 0x30, // TODO: Implement support
 
   Void = 0xff,
 }
 
+export type ContractParamMap = {
```
Comment (snowypowers) [inline packages/neon-core/src/sc/ContractParam.ts:26]: Propose to rename this to `ContractParamMapEntry`.  `ContractParamMap` is too close to the type `ContractParam.Map` and causes some confusion.

File: packages/neon-core/src/sc/ContractParam.ts [hunk 5]
```diff
@@ -16,22 +16,44 @@ export enum ContractParamType {
...
+
 export interface ContractParamJson {
   type: string;
-  value?: string | boolean | number | ContractParamJson[] | null;
+  value?:
+    | string
+    | boolean
+    | number
+    | ContractParamJson[]
+    | ContractParamMapJson
```
Comment (snowypowers) [inline packages/neon-core/src/sc/ContractParam.ts:48]: I believe this should be `ContractParamMapJson[]` (or `ContractParamMapEntryJson[]` if renamed)

File: packages/neon-core/src/sc/ContractParam.ts [hunk 6]
```diff
@@ -16,22 +16,44 @@ export enum ContractParamType {
...
-  Map = 0x22, // TODO: Implement support
+  Map = 0x22,
 
   InteropInterface = 0x30, // TODO: Implement support
 
   Void = 0xff,
 }
 
+export type ContractParamMap = {
+  key: ContractParam;
```
Comment (snowypowers) [inline packages/neon-core/src/sc/ContractParam.ts:27]: Do add a comment here to constraint the type.

**Review by snowypowers — APPROVED** [review]

**Pull request merged**